pkg/controllermanager/controller/plant/plant_health.go

```diff
@@ -0,0 +1,123 @@
+package plant
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"k8s.io/client-go/discovery"
+
+	"github.com/gardener/gardener/pkg/utils/kubernetes/health"
+
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"k8s.io/apimachinery/pkg/labels"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+)
+
+// NewHealthCheker creates a new health checker.
+func NewHealthCheker(plantClient client.Client, discoveryClient discovery.DiscoveryInterface) *HealthChecker {
+	return &HealthChecker{
+		plantClient:     plantClient,
+		discoveryClient: discoveryClient,
+	}
+}
+
+// HealthChecker contains the condition thresholds.
+type HealthChecker struct {
+	plantClient     client.Client
+	discoveryClient discovery.DiscoveryInterface
+}
+
+// CheckPlantClusterNodes checks whether cluster nodes in the given listers are complete and healthy.
+func (h *HealthChecker) CheckPlantClusterNodes(condition *gardencorev1alpha1.Condition, nodeLister kutil.NodeLister) (*gardencorev1alpha1.Condition, error) {
+	nodeList, err := nodeLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	if exitCondition := h.checkNodes(*condition, nodeList); exitCondition != nil {
+		return exitCondition, nil
+	}
+
+	updatedCondition := helper.UpdatedCondition(*condition, corev1.ConditionTrue, "EveryNodeReady", "Every node registered to the cluster is ready")
+
+	return &updatedCondition, nil
+}
+
+// CheckAPIServerAvailability checks if the API server of a Shoot cluster is reachable and measure the response time.
+func (h *HealthChecker) CheckAPIServerAvailability(condition gardencorev1alpha1.Condition) gardencorev1alpha1.Condition {
+	// Try to reach the Shoot API server and measure the response time.
+	now := time.Now()
+	discoveryClient, ok := h.discoveryClient.(discovery.DiscoveryInterface)
```

Comment:
hmm, I think it is still better to have two seperate implementation for the following reasons: 

- The client used is different, we are directly using the kubernetes discovery bare client and not our whole gardener's Kubernetes Interface (we only need discovery. 
- We don't have a progressing state so we are not using `FailedCondition` which relies on condition thresholds. 
- We are using `gardencorev1alpha1.Condition` and not `gardenv1beta1.Condition`.

The only thing left to re-use is the status checking which imho makes it not worth the effort, wdyt?

Reply:
> The client used is different, we are directly using the kubernetes discovery bare client and not our whole gardener's Kubernetes Interface (we only need discovery).

That could be harmonized?

> We don't have a progressing state so we are not using FailedCondition which relies on condition thresholds.

That can be wrapped to prevent code duplication

> We are using gardencorev1alpha1.Condition and not gardenv1beta1.Condition.

With #792 there will only be `gardencorev1alpha1.Condition`, `gardenv1beta1.Condition` will be removed entirely

Reply:
Okay, then I would suggest to wait till #792 is merged and then we refactor this function properly.